src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -203,15 +196,78 @@ void NnetDecodableBase::GetCurrentIvector(int32 output_t_start,
                 << " (mismatched --ivector-period?)";
     }
     ivector_frame = online_ivector_feats_->NumRows() - 1;
+  } else if (ivector_frame < 0) {
+    ivector_frame = 0;
+  }
+  ivector->CopyFromVec(online_ivector_feats_->Row(ivector_frame));
+}
+
+void NnetDecodableBase::GetIvectorsForFrames(int32 output_t_start,
+                                             int32 num_output_frames,
+                                             Matrix<BaseFloat> *ivectors) {
+  const int32 left_context = nnet_left_context_ + opts_.extra_left_context,
+        right_context = nnet_right_context_ + opts_.extra_right_context;
+  // frame_to_search is the frame that we want to get the most recent iVector
+  // for.  We choose a point near the middle of the current window, the concept
+  // being that this is the fairest comparison to nnet2.   Obviously we could do
+  // better by always taking the last frame's iVector, but decoding with
+  // 'online' ivectors is only really a mechanism to simulate online operation.
+  if (ivector_interval_ == 0) { // single ivector case
+    ivectors->Resize(1, online_ivector_feats_->NumCols(), kUndefined);
+    int32 frame_to_search = output_t_start + num_output_frames / 2;
+    SubVector<BaseFloat> sub_ivector = ivectors->Row(0);
+    GetCurrentIvector(frame_to_search, &sub_ivector);
+  } else { // multiple ivectors case
+    // num_ivectors_in_left_context is the num of ivectors for frames
+    // whose "t" index < 0.
+    // num_ivectors is the num of ivectors for the entire chunk.
+    // They are computed according to how Round descriptor works, which
+    // basically returns floor(t / <t-modulus>) * <t-modulus>.
+    // (The assumption is that the t index of the first outptut frame is 0)
+    int32 num_ivectors_in_left_context =
+        std::ceil(1.0 * left_context / ivector_interval_);
+    int32 num_ivectors = num_ivectors_in_left_context +
+        (num_output_frames + right_context - 1) / ivector_interval_ + 1;
+    ivectors->Resize(num_ivectors, online_ivector_feats_->NumCols(), kUndefined);
+    // the position the right-most ivector in left context is at the middle of
+    // the interval, which is computed as
+    // output_t_start - 1 - std::min(ivector_interval_, left_context) / 2.
+    // Then we compute the left-most ivector to start as frames_to_search 
+    int32 frame_to_search = output_t_start - 1 -
+        std::min(ivector_interval_, left_context) / 2 -
+        (num_ivectors_in_left_context - 1) * ivector_interval_;
+    for (int32 n = 0; n < num_ivectors; n++) {
+      if (n == num_ivectors_in_left_context)
+        // the position of the left-most ivector when t>=0 is at the middle of
+        // the interval
+        frame_to_search = output_t_start +
+                          std::min(ivector_interval_, num_output_frames) / 2;
```

Comment:
This whole block is extremely strange, IMO.  We just want the iVectors at
whole multiples of the ivector_interval, the left/right distinction is
unnecessary.

Dan


On Fri, Mar 11, 2016 at 12:56 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-am-decodable-simple.cc
> <https://github.com/kaldi-asr/kaldi/pull/468#discussion_r55863791>:
>
> > +    int32 num_ivectors = num_ivectors_in_left_context +
> > +        (num_output_frames + right_context - 1) / ivector_interval_ + 1;
> > +    ivectors->Resize(num_ivectors, online_ivector_feats_->NumCols(), kUndefined);
> > +    // the position the right-most ivector in left context is at the middle of
> > +    // the interval, which is computed as
> > +    // output_t_start - 1 - std::min(ivector_interval_, left_context) / 2.
> > +    // Then we compute the left-most ivector to start as frames_to_search
> > +    int32 frame_to_search = output_t_start - 1 -
> > +        std::min(ivector_interval_, left_context) / 2 -
> > +        (num_ivectors_in_left_context - 1) * ivector_interval_;
> > +    for (int32 n = 0; n < num_ivectors; n++) {
> > +      if (n == num_ivectors_in_left_context)
> > +        // the position of the left-most ivector when t>=0 is at the middle of
> > +        // the interval
> > +        frame_to_search = output_t_start +
> > +                          std::min(ivector_interval_, num_output_frames) / 2;
>
> I thought we always ensured ivector_interval_ is always less than
> chunk_width. So why do you need the min ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/468/files#r55863791>.
>